stateMutators.js

```diff
@@ -0,0 +1,174 @@
+const gameState = {
+    basketPos: 50,
+    basketValue: 0,
+    score: 0,
+    livesRemaining: 3,
+    gameLevel: 1,
+    gameMode: 'running',
+    fallingObjects: [],
+}
+
+const setBasketPos = (state, value) => {
+    return {
+        ...state,
+        basketPos: value,
+    }
+};
+
+const moveBasketLeft = (state, value) => {
+    return {
+        ...state,
+    basketPos: basketPos - value,
+    }
+};
+
+const moveBasketRight = (state, value) => {
+    return {
+        ...state,
+        basketPos: basketPos + value,
+    }
+};
+
+const setBasketValue = (state, value) => {
+    return {
+        ...state,
+        basketValue: value,
+    }
+};
+
+const resetBasketValue = (state) => {
+    return {
+        ...state,
+        basketValue: 0,
+    }
+};
+
+const setScore = (state, value) => {
+    return {
+        ...state,
+        score: value,
+    }
+};
+
+const resetScore = (state) => {
+    return {
+        ...state,
+        score: 0,
+    }
+};
+
+const loseLife = (state, value) => {
+    return {
+        ...state,
+        livesRemaining: value,
+    }
+};
+
+const resetLivesRemaining = (state) => {
+    return {
+        ...state,
+        livesRemaining: 3,
+    }
+};
+
+const setGameLevel = (state, value) => {
+    return {
+        ...state,
+        gameLevel: value,
+    }
+};
+
+const resetGameLevel = (state) => {
+    return {
+        ...state,
+        gameLevel: 0,
+    }
+};
+
+const setGameMode = (state, value) => {
+    return {
+        ...state,
+        gameMode: value,
+    }
+};
+
+const addFallingObject = (state, fallingObj) => {
+    return {
+        ...state,
+        fallingObjects: [
+            ...state.fallingObjects,
+            fallingObj,
+        ]
+    }
+};
+
+const removeFallingObject = (state, fallingObjId) => {
+    return {
+        ...state,
+        fallingObjects: [
+            ...state.fallingObjects.slice[0, fallingObjId],
+            ...state.fallingObjects.slice[fallingObjId + 1],
+        ]
+    }
+};
+
+
+// Object constructor for falling object
+function FallingObject(posX, posY, numerator, denominator) {
+    this.posX = posX;
+    this.posY = posY;
+    this.numerator = numerator;
+    this.denominator = denominator;
+    this.value = numerator / denominator;
+    this.id = id;
+}
+
+const setFallingObjPosY = (fallingObj, value) => {
+    fallingObj.posY = value;
+}
+
+
+const setFallingObjNumerator = (fallingObj, value) => {
+    fallingObj.numerator = value;
+}
+
+const setFallingObjDenominator = (fallingObj, value) => {
+    fallingObj.denominator = value;
+}
+
+// Calculating Functions
+
+const calcBasketValue = (state, basketValue, fallingObjectValue) => {
+    return {
+        ...state,
+        basketValue: basketValue + fallingObjectValue,
+    }
+};
+
+
+const calcScore = (state, basketValue, score) => (
```

Comment:
Same as the above with the 2nd and 3rd params. They can be derived from the first `state`.

I also would prefer `score: state.score + 1,`, or adding a new param you can pass in `score: state.score + scoreIncrease,`, rather than using `++` here though. This particular usage should work fine, but using `++` when setting things can easily lead to bugs as it also modifies the underlying variable - which lead to unexpected values in my experience.
Same with the `--livesRemaining` below

Reply:
roger that